File: GoLang/solution/solution0404.go
package solution

import (
    "gproject/dataStructure"
)

/*
计算给定二叉树的所有左叶子之和。
示例：
    3
   / \
  9  20
    /  \
   15   7
在这个二叉树中，有两个左叶子，分别是 9 和 15，所以返回 24
*/
// type TreeNode struct {
// 	Val int
// 	Left *TreeNode
// 	Right *TreeNode
// }

func SumOfLeftLeaves(root *datastructure.TreeNode) int {
    sum := 0
    preOrder(&sum, root)
    return sum
}

func preOrder(sum *int, node *datastructure.TreeNode) {
    if node == nil {
        return
    } else {
        if node.Left != nil && node.Left.Left == nil && node.Left.Right == nil {
			*sum = *sum + node.Left.Val
		}
	}
	preOrder(sum, node.Left)
	preOrder(sum, node.Right)
}

File: JavaR2/src/Solution0404.java
/**
 * @author maximusyoung
 */
public class Solution0404 {
	int res = 0;
	public int sumOfLeftLeaves(TreeNode root) {
		cal(root, false);
		return res;
	}

	public void cal(TreeNode root, boolean isLeft) {
		if (root == null) {
			return;
		}

		if (isLeft && root.left == null && root.right == null) {
			res += root.val;
		}

		cal(root.left, true);
		cal(root.right, false);

	}
}
